### PR TITLE
add onChange handler to next/Stepper

### PR DESCRIPTION
**Pull request by dagda1**
## Hey, I just made a Pull Request!

We have some templates that contain lots of placeholders and `ui:field` extensions and they were not initialised correctly.

The `formData`  that is passed to all the extensions and placeholders is not in sync with the `@rjsf` `<Form />` component.

The current `MultistepJson` form component uses an `onChange` handler to keep the `formData` in sync with the `<Form />` component.

This PR adds an `onChange` handler to the `<Stepper/>` component and seems to do an excellent job of keeping the `formData` in sync.  The `getDefaultFormState` that I added in #14660 is no longer needed with this approach, and I have removed it.

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx [hunk 1]
```diff
@@ -102,6 +101,11 @@ export const Stepper = (props: StepperProps) => {
     setActiveStep(prevActiveStep => prevActiveStep - 1);
   };
 
+  const handleChange = useCallback(
+    (e: IChangeEvent) => setFormState(e.formData),
```
Comment (benjdlambert) [inline plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx:105]: If we have multiple steps - we don't end up replacing all previous data too do we, or do we need to merge here like we do on https://github.com/backstage/backstage/pull/15402/files#diff-b634e1da6a661c7e4d8e1205c15d5392edeb09c6eb37595cf4b0705d2e051b45R134

Comment (dagda1) [reply plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx:105]: @benjdlambert we don't, I'll add a good test to prove this

Comment (dagda1) [reply plugins/scaffolder/src/next/TemplateWizardPage/Stepper/Stepper.tsx]: @benjdlambert that is actually a good shout to merge.  I have updated to use the merge

**Comment by dagda1** [comment]
@benjdlambert I've added a test to check the `formData` is merging correctly

**Review by benjdlambert — APPROVED** [review]
Nice work! Looks good!

**Pull request merged**